feat: metamask notifications controller

matteoscurati opened the pull request:
## **Description**

This PR introduces a new controller for managing two different types of notifications:

- feature announcements
- notifications related to events and actions for the user-managed accounts

These notifications are received from two external sources.

The new controller is called `Metamask Notifications Controller` to distinguish it from the current `Notification Controller` used to manage the internal notifications of the extension.

The `Metamask Notifications Controller` requires two controllers for managing and receiving notifications:

- `authorization controller` - responsible for "authenticating" the user
- `user storage controller` - responsible for the creation and reading of the user storage

With the combination of using these two controllers, the user can choose whether and which notifications to receive.

## **Related issues**

Fixes:

## **Manual testing steps**

## **Screenshots/Recordings**

Check this Figma to have an overview about  the flow of the Platform Notifications Controller:

[Figma - Notifications](https://www.figma.com/file/c7GgNw2kScGrVyRGAPhwEd/Notifications?type=design&node-id=484-44713&mode=design)

### **Before**

### **After**

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

matteoscurati committed:
feat: :sparkles: metamask notifications controller

matteoscurati committed:
Merge branch 'develop' into feature/metamask-notifications-controller

matteoscurati committed:
refactor: :recycle: remove unused logic and update the controller

matteoscurati committed:
refactor: :label: typo

matteoscurati committed:
refactor: :bug: fix the import of the feature announcement notifications

matteoscurati committed:
feat: :sparkles: avoit to fetch notifications if the notifications are not enabled

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

Prithpal-Sooriya committed:
refactor: cleanup

fix eslint errors, type errors, import errors. Also remove shared constants as we will be moving this into a shared library in future.

Prithpal-Sooriya committed:
test: Add endpoint services tests

Prithpal-Sooriya committed:
refactor: mark check trigger groups as deprecated

discuss with team to remove this initial implementation

Prithpal-Sooriya committed:
test: util tests

Prithpal-Sooriya committed:
test: test processors

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

Prithpal-Sooriya committed:
refactor: cleanup notification controller

update methods, using messaging system, cleanup constructor

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -156,134 +208,88 @@ export class MetamaskNotificationsController extends BaseController<
+  #metamaskNotificationUtils: MetamaskNotificationsUtils;
 
-  private updatePushNotifications: (UUIDs: string[]) => Promise<void>;
+  #auth = {
```
All are external controllers now go through the messaging system.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -376,14 +380,43 @@ export class MetamaskNotificationsController extends BaseController<
     );
   }
 
+  #assertAuthEnabled() {
```
These are some utils to get and assert User Storage; Bearer Token; Auth; etc.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -428,10 +461,11 @@ export class MetamaskNotificationsController extends BaseController<
   public async checkTriggersPresenceByGroup(): Promise<
```
I've kept this method, but lets not use this yet. I'm not confident in our implementation of group trigger enable/disable.

Doing an introspection and "guessing" if a filter should be enabled/disabled is really hard, especially breaks down when you also need to consider any new notifications we will add in the future.

Proposal: It might be easier to have a new User Storage entry to say if a filter is enabled/disabled instead of introspecting.

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

Prithpal-Sooriya committed:
test: initial controller tests

tested controller and some basic methods. Need to test more in-depth methods like creating notifications and logging in. But this is hard since the classes make it very difficult to mock.

Prithpal-Sooriya committed:
refactor: convert classes into modules

the services and utils should be modules, not classes. This makes it easier for tree shaking and mocking too (I can mock 1 function instead of the whole class)

Prithpal-Sooriya committed:
refactor: tidy up notification accounts cache and add some more tests

just a few more functionality that needs to be tested

Prithpal-Sooriya committed:
test: add notification controller tests

test all public method, subscriptions and control flow paths

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

Prithpal-Sooriya commented on app/scripts/controllers/authentication/authentication-controller.ts:
```diff
@@ -152,6 +157,10 @@ export default class AuthenticationController extends BaseController<
     return profile;
   }
 
+  public isSignedIn(): boolean {
+    return this.state.isSignedIn;
+  }
```
This is duplicate from an earlier PR. Will be removed once User Storage PR is merged.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/constants/notification-schema.ts:
```diff
@@ -0,0 +1,163 @@
+  DEFI = 'defi',
+}
+
+export const NOTIFICATION_CHAINS = {
+  ETHEREUM: '1',
+  OPTIMISM: '10',
+  BSC: '56',
+  POLYGON: '137',
+  ARBITRUM: '42161',
+  AVALANCHE: '43114',
+  LINEA: '59144',
+};
```
We are keeping a local copy of the chains here as we need this controller to be loosely coupled with the extension (it will need to be reused with other platforms).

If there is a nice reusable list of chains from a shared package that would be perfect!

danjm replied:
We do not have a reusable list of chains in a shared package at this time.

Prithpal-Sooriya replied:
no worries that is fine!

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/constants/notification-schema.ts:
```diff
@@ -0,0 +1,163 @@
+  },
+};
+
+export enum TRIGGER_DATA_TYPES {
+  DATA_FEATURE_ANNOUNCEMENT = 'data_feature_announcement',
+  DATA_METAMASK_SWAP_COMPLETED = 'data_metamask_swap_completed',
+  DATA_STAKE = 'data_stake',
+  DATA_LIDO_STAKE_READY_TO_BE_WITHDRAWN = 'data_lido_stake_ready_to_be_withdrawn',
+  DATA_ETH = 'data_eth',
+  DATA_ERC20 = 'data_erc20',
+  DATA_ERC721 = 'data_erc721',
+  DATA_ERC1155 = 'data_erc1155',
+}
```
Removing this, we don't need it

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/encryption/encryption.ts:
```diff
@@ -0,0 +1,7 @@
+import { sha256 } from '@noble/hashes/sha256';
+import { bytesToHex } from '@noble/hashes/utils';
+
+export function createSHA256Hash(data: string): string {
+  const hashedData = sha256(data);
+  return bytesToHex(hashedData);
+}
```
This is temporary duplication. Will be removed when User Storage is merged.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -0,0 +1,955 @@
+    // Subscription when an new account is changed
+    // The initial cache should already have addresses already have in wallet,
+    // so any new account switched to (that is not in cache) We can assume it is new.
+    createNewNotificationsOnNewAccountChange: () => {
+      this.messagingSystem.subscribe(
+        'AccountsController:selectedAccountChange',
```
As the comment suggested - we had to use this event for figuring out if new addresses were added to the wallet.

Is there are more reliable event we can listen to?

danjm replied:
Is there is a concern about reliability here? Does `AccountsController:selectedAccountChange` not occur when you would expect it to?

danjm replied:
I wonder how it behaves for adding multiple accounts during hardware wallet connection flows :thinking:

danjm replied:
Also, we may want to handle the case where hardware wallet accounts, or snap accounts, are removed.

Prithpal-Sooriya replied:
The `AccountsController:selectedAccountChange` does work as we intended (so far) - but was just wondering if there is a specific event just for adding new accounts instead of when an account is changed.

I'll do some additional testing to ensure we are targeting the correct accounts - great callout!

Prithpal-Sooriya replied:
> accounts removed

Great callout! Will ned to check this.

Prithpal-Sooriya replied:
If this is a bit more challenging than expected then we can have this in another PR

Prithpal-Sooriya replied:
Instead of using the AccountsController, we've opted for the Keyring Controller does a pretty good job for our use-case!

Also covered multi-add or multi-remove of accounts.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/services/feature-announcements.ts:
```diff
@@ -0,0 +1,129 @@
+): Promise<ContentfulResult | null> {
+  let lastError: Error | null = null;
+
+  for (let i = 0; i < retries; i++) {
+    try {
+      const response = await fetch(url);
+      if (!response.ok) {
+        throw new Error(`Fetch failed with status: ${response.status}`);
+      }
+      return await response.json();
+    } catch (error) {
+      if (error instanceof Error) {
+        lastError = error;
+      }
+      if (i < retries - 1) {
+        await new Promise((resolve) => setTimeout(resolve, retryDelay));
+      }
+    }
+  }
+
```
Hmm iffy about retries on controller logic (IDK if retry logic should be done at this level or UI level).

Lets keep this  here since you've implemented it, but once we start connecting the UI lets sync and cleanup if necessary.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/services/onchain-notifications.ts:
```diff
@@ -0,0 +1,288 @@
+  }
+
+  const onChainNotifications: OnChainRawNotification[] = [];
+  const PAGE_LIMIT = 2;
```
OK IDM pagination at controller level. We can tweak this once we connect the UI.

If a page = 100 on-chain notifications then we will have:
```
+ 200 On-Chain
+ X Off-Chain (aka feature announcement)
+ Y Snap notifications
---
Alot of notifications
```

It might not be reasonable to show this many notifications. Lets update once we integrate UI.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/types/on-chain-notification/on-chain-notification.ts:
```diff
@@ -0,0 +1,50 @@
+ * 2. It ensures that the `data` field is the correct Notification data for this `type`
+ * - The `Compute` utility merges the intersections (`&`) for a prettier type.
+ */
+export type OnChainRawNotification = {
```
This is internal types. Other users will not need to manage to handle these.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/types/types.ts:
```diff
@@ -0,0 +1,27 @@
+>;
+
+// NFT
+export type NFT = {
+  token_id: string;
+  image: string;
+  collection?: {
+    name: string;
+    image: string;
+  };
+};
```
Not used - remove

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/types/user-storage/user-storage.ts:
```diff
@@ -0,0 +1,47 @@
+ * The User Storage Endpoint requires a path and an entry name.
+ * Developers can provide additional paths by extending this variable below
+ */
+export const USER_STORAGE_ENTRIES = {
+  notification_settings: {
+    path: 'notifications',
+    entryName: 'notification_settings',
+  },
+} satisfies Record<string, UserStorageEntry>;
+
+export type UserStorageEntryKeys = keyof typeof USER_STORAGE_ENTRIES;
```
This is declared in the User Storage PR - will clean up once it is merged.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/utils/utils.ts:
```diff
@@ -0,0 +1,621 @@
+import log from 'loglevel';
```
This utils is huge. Mostly because was a class.

I've refactored this to instead use modules (makes it much more easier to test!)

FUTURE - lets separate and group utils into different files. E.g. User Storage utils (traversing, mapping, etc); Fetching logic; User Storage mutations.

danjm replied:
:+1:

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/utils/utils.ts:
```diff
@@ -0,0 +1,621 @@
+ * @returns A Promise resolving to the Response object.
+ * @throws Will throw an error if the request fails after the specified number of retries.
+ */
+async function fetchWithRetry(
```
See prev comment on retries. IDK if this should live in controller level or on UI level.

It feels weird to re-invent the wheel here when RTK-Query or other data/state management libs can handle this.

Lets refactor this in a future PR once we connect and play with the UI.

Prithpal-Sooriya commented on app/scripts/metamask-controller.js:
```diff
@@ -217,6 +217,7 @@ import { STATIC_MAINNET_TOKEN_LIST } from '../../shared/constants/tokens';
 import { getTokenValueParam } from '../../shared/lib/metamask-controller-utils';
 import { isManifestV3 } from '../../shared/modules/mv3.utils';
 import { convertNetworkId } from '../../shared/modules/network.utils';
+
```
This file doesn't need changes... remove whitespace changes.

Lets connect this controller in the UI integration PRs.

Prithpal-Sooriya commented on builds.yml:
```diff
@@ -29,6 +29,12 @@ buildTypes:
       - REQUIRE_SNAPS_ALLOWLIST: true
       - IFRAME_EXECUTION_ENVIRONMENT_URL: https://execution.metamask.io/iframe/6.0.1/index.html
       - ACCOUNT_SNAPS_DIRECTORY_URL: https://snaps.metamask.io/account-management
+      - CONTENTFUL_ACCESS_SPACE_ID: null
+      - CONTENTFUL_ACCESS_TOKEN: null
+      - AUTH_LOGIN_ENDPOINT: null
+      - USER_STORAGE_ENDPOINT: null
+      - NOTIFICATIONS_SERVICE_URL: null
+      - TRIGGERS_SERVICE_URL: null
```
We are not just targeting main but all builds so will move this.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/processors/process-feature-announcement.ts:
```diff
@@ -0,0 +1,21 @@
+  notificationId: string,
+  readPlatformNotificationsList: string[],
+): boolean {
+  return readPlatformNotificationsList.includes(notificationId);
```
Lets add the 1-month stale period for feature announcements (there is no point in showing a notification that is >1 month old). We can tweak if need to.

Prithpal-Sooriya committed:
refactor: self-review changes

tidy up comments, remove unused types, and move envs.

Prithpal-Sooriya committed:
fix: fix linting

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

danjm commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -0,0 +1,954 @@
+   *
+   * @async
+   * @throws {Error} If updating the state fails.
+   */
```
If notifications are toggled off, and the user adds new accounts, and then notifications are toggled on, do we need to do something like `updateOnChainTriggersByAccount` for those accounts?

Prithpal-Sooriya replied:
When notifications are toggled on we'll call `createOnChainTriggers()` which will go through all accounts and subscribe them to notifications - so it should cover any new accounts that were added.

We can probably add some smarter logic in the future - e.g. saving users preferences to have accounts that were turned off previously stay turned off.

Prithpal-Sooriya replied:
Note this is def hard for an external reviewer to review - since this controller is kind-of decoupled from the UI so it is hard to get some context.

I'll make sure that these public methods are well documented and give an idea of where they are used.

Prithpal-Sooriya replied:
I've updated the JSDoc comments here to include an `**Action**` on the intent of each method.

danjm commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -0,0 +1,954 @@
+
+          // An actual new address! We can create triggers/notifications for this.
+          this.#accountsCache.add(checksumAddress);
+          await this.updateOnChainTriggersByAccount(checksumAddress);
```
Does this also need to be called for the accounts cached in the `listAccounts` method?

Prithpal-Sooriya replied:
This has now been covered, we do make sure that we support multi-add or remove accounts.

danjm commented on app/scripts/controllers/metamask-notifications/types/on-chain-notification/on-chain-notification.ts:
```diff
@@ -0,0 +1,50 @@
+}[TRIGGER_TYPES];
+
+/**
+ * Type-Foo.
```
Does `Type-Foo` need to be updated?

Prithpal-Sooriya replied:
LOL I meant `Type-Fu` (like kung fu) 😛 

I'll rename this comment

danjm commented on app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts:
```diff
@@ -0,0 +1,715 @@
+});
+
+describe('metamask-notifications - deleteOnChainTriggersByAccount', () => {
+  test('Deletes and disabled push notifications for a given account', async () => {
```
Nit, typo, "disabled" -> "disables"

Prithpal-Sooriya replied:
Lol yep!

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/services/feature-announcements.ts:
```diff
@@ -0,0 +1,129 @@
+
+const spaceId = process.env.CONTENTFUL_ACCESS_SPACE_ID || '';
+const accessToken = process.env.CONTENTFUL_ACCESS_TOKEN || '';
+export const FEATURE_ANNOUNCEMENT_URL = `https://cdn.contentful.com/spaces/${spaceId}/environments/master/entries?access_token=${accessToken}&content_type=productAnnouncement&include=10`;
```
Lets put this under env/build.

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -0,0 +1,955 @@
+  metamaskNotificationsReadList: [],
+};
+
+// TODO - Mock Storage Controller Actions, added in a separate PR.
```
USer storage is merged!!

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feature/metamask-notifications-controller

Prithpal-Sooriya committed:
refactor: requested changes - tidy up

improve comments, add better auto sub/unsub logic, remove duplicate code

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/metamask-notifications.ts:
```diff
@@ -0,0 +1,833 @@
+     */
+    subscribe: () => {
+      this.messagingSystem.subscribe(
+        'KeyringController:stateChange',
```
Sweet - looks around and I think the keyring controller events/actions are better suited for our use case. Tested & seems to work well!

- It targets wallet specific addresses that we want to track (not snap or other accounts)

- We can use the state change event to then check when new addresses are added or removed (see logic below)
  - there is a account removed event, but no account added event. But the below logic works nicely!

danjm replied:
yeah, that makes sense! good find

Prithpal-Sooriya committed:
docs: jsdoc comment

Prithpal-Sooriya committed:
test: fix failing test

this was a edge case issue where the test actually needed to return true

Prithpal-Sooriya commented on app/scripts/controllers/metamask-notifications/constants/notification-schema.ts:
```diff
@@ -0,0 +1,152 @@
+  TRIGGER_TYPES.FEATURES_ANNOUNCEMENT
+>;
+
+export const TRIGGERS: Record<OnChainTriggerTypes, Trigger> = {
```
Notifications Dev Note (aka note for self) - for future notifications we will support, we need to update this schema. The system will automatically add any new notifications for users.

In future (next quarter) once we add NFD we'll probably have a single source of truth for this schema shared client and server side.

Prithpal-Sooriya commented:
NOTE - this controller is not yet connected to the main `metamask-controller.js` file. @matteoscurati when push notifications are integrated or do the UI integration can you then plug in the controller?
